internal/webhook/sentinel/handler.go

```diff
@@ -0,0 +1,196 @@
+	// Allow updates to such leases, but only by etcd members, which would use the serviceaccount deployed by druid for them.
+	if requestGK == coordinationv1.SchemeGroupVersion.WithKind("Lease").GroupKind() &&
+		req.Operation == admissionv1.Update {
+		saTokens := strings.Split(req.UserInfo.Username, ":")
+		saName := saTokens[len(saTokens)-1] // last element of sa name which looks like `system:serviceaccount:<namespace>:<name>`
+		if saName == etcd.GetServiceAccountName() {
+			return admission.Allowed("lease resource can be freely updated by etcd members")
+		}
+	}
+
+	// allow changes to resources if Etcd has annotation druid.gardener.cloud/disable-resource-protection is set
+	if !etcd.AreManagedResourcesProtected() {
+		return admission.Allowed(fmt.Sprintf("changes allowed, since Etcd %s has annotation %s", etcd.Name, druidv1alpha1.DisableResourceProtectionAnnotation))
+	}
+
+	// allow operations on resources if the Etcd is currently being reconciled, but only by etcd-druid,
+	if etcd.IsReconciliationInProgress() {
+		if req.UserInfo.Username == h.config.ReconcilerServiceAccount {
+			return admission.Allowed(fmt.Sprintf("ongoing reconciliation of Etcd %s by etcd-druid requires changes to resources", etcd.Name))
+		}
+		return admission.Denied(fmt.Sprintf("no external intervention allowed during ongoing reconciliation of Etcd %s by etcd-druid", etcd.Name))
+	}
+
+	// allow exempt service accounts to make changes to resources, but only if the Etcd is not currently being reconciled.
+	for _, sa := range h.config.ExemptServiceAccounts {
+		if req.UserInfo.Username == sa {
+			return admission.Allowed(fmt.Sprintf("operations on Etcd %s by service account %s is exempt from Sentinel Webhook checks", etcd.Name, sa))
+		}
+	}
+
+	return admission.Denied(fmt.Sprintf("changes disallowed, since no ongoing processing of Etcd %s by etcd-druid", etcd.Name))
+}
+
+// decodeRequestObject decodes the relevant object from the admission request and returns it.
+// If it encounters an unexpected resource type, it returns a nil object.
+func (h *Handler) decodeRequestObject(req admission.Request, requestGK schema.GroupKind) (client.Object, error) {
+	var (
+		obj client.Object
+		err error
+	)
+	switch requestGK {
+	case
+		corev1.SchemeGroupVersion.WithKind("ServiceAccount").GroupKind(),
+		corev1.SchemeGroupVersion.WithKind("Service").GroupKind(),
+		corev1.SchemeGroupVersion.WithKind("ConfigMap").GroupKind(),
+		rbacv1.SchemeGroupVersion.WithKind("Role").GroupKind(),
+		rbacv1.SchemeGroupVersion.WithKind("RoleBinding").GroupKind(),
+		appsv1.SchemeGroupVersion.WithKind("StatefulSet").GroupKind(),
+		autoscalingv1.SchemeGroupVersion.WithKind("Scale").GroupKind(), // for statefulsets' /scale subresource
+		policyv1.SchemeGroupVersion.WithKind("PodDisruptionBudget").GroupKind(),
+		batchv1.SchemeGroupVersion.WithKind("Job").GroupKind(),
+		coordinationv1.SchemeGroupVersion.WithKind("Lease").GroupKind():
```

Comment:
This should be in one common place soemthing which combines the usage we defined r.operatorRegistry.AllOperators() and here. So that both refer to same source of truth maintained at one place and consumed as per the caller needs.
Rationale, is if one wants to introduce a new operator/kind managed by druid as part of ETCD resource, one needs to remember that the webhook also needs to be updated as well.

Reply:
Sure, it makes sense now to me.